src/workerd/util/entropy-source.c++

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2017-2022 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+#include "entropy-source.h"
+
+namespace workerd {
+
+void BufferedEntropySource::generate(kj::ArrayPtr<kj::byte> buffer) {
+  thread_local BufferState state;
+
+  while (buffer != nullptr) {
+    if (state.data == nullptr) {
+      refillBuffer(state);
+    }
+
+    size_t toCopy = kj::min(state.data.size(), buffer.size());
+    buffer.first(toCopy).copyFrom(state.data.first(toCopy));
+    state.data = state.data.slice(toCopy);
+    buffer = buffer.slice(toCopy);
+  }
+}
+
+void BufferedEntropySource::refillBuffer(BufferState& state) {
+  KJ_ASSERT(RAND_bytes(state.store.begin(), BUFFER_SIZE) == 1,
+      "RAND_bytes failed to generate random data");
+  state.data = state.store.asPtr();
+}
+
+void BufferedEntropySource::generateBatched(
+    kj::ArrayPtr<uint64_t> out, kj::Maybe<kj::EntropySource&> entropySource) {
+  KJ_IF_SOME(entropy, entropySource) {
+    entropy.generate(
+        kj::arrayPtr(reinterpret_cast<kj::byte*>(out.begin()), out.size() * sizeof(uint64_t)));
+  } else {
+    KJ_ASSERT(
+        RAND_bytes(reinterpret_cast<uint8_t*>(out.begin()), out.size() * sizeof(uint64_t)) == 1);
```

Comment:
Same here... the crypto error queue needs to be emptied on return. There's an ncrypto utility for doing that.